Add a subdomain for certificates.

aweisser opened the pull request:
A "Certificate" is a uniquely identifiable and immutable struct
which certifies that the owner of the certificate has executed a
certain "Activity".

All attributes of a "Certificate" are semi final, means they are immutable after beeing set.
All attributes are also package private, so if you want to create a "Certificate"
you have to use the exportet factory functions of this package.

@davelosert  please have a look.

davelosert committed:
Created a first, very simple make file for the rest-server

davelosert committed:
Added dependency management and testing to makefile

davelosert committed:
Added clean command

davelosert committed:
Removed windows build and added linting

davelosert committed:
Added metalinter install command so it's deps get installed

davelosert committed:
Excluded generated directories from linting

davelosert committed:
Fixed linting error: Too long comment in main.go

davelosert committed:
Fixed linting error: Shadowed member-variable

aweisser committed:
Add a subdomain for certificates.

A "Certificate" is a uniquely identifiable and immutable struct
which certifies that the owner of the certificate has executed a
certain "Activity".

aweisser committed:
Add a subdomain for certificates.

A "Certificate" is a uniquely identifiable and immutable struct
which certifies that the owner of the certificate has executed a
certain "Activity".

davelosert commented on domain/cert/cert_test.go:
```diff
@@ -0,0 +1,150 @@
+				So(err, ShouldNotBeNil)
+			})
+		})
+	})
```
Isn't this the same test twice?
Please delete the duplicated one.

davelosert commented on domain/cert/certificate.go:
```diff
@@ -0,0 +1,82 @@
+// ID is a globally unique ID. Perhaps a UUID?
+type ID interface{}
+
+// A Certificate certifies an Activity and belongs to an owner (identified by any id)
```
This comment is duplication - the property `ownerID` speaks for itself.

If the certificate is JUST for activities, instead of commenting please rather rename it to `ActivityCertificate` to be more clear.

davelosert commented on domain/cert/certificate.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// AssignTo assigns the Certificate to an owner. This can only be executed ONCE!
+// So a Certificate can only be assigned once.
```
Also remove this comment - rather unnecessary, the code is clean enough.

davelosert commented on domain/cert/certificate.go:
```diff
@@ -0,0 +1,82 @@
+	category string // should be set by authorized cert Providers (inside this package)
+}
+
+// Category of this Activity
```
... another unnecessary comment.

davelosert commented on domain/cert/certificate.go:
```diff
@@ -0,0 +1,82 @@
+	"github.com/satori/go.uuid"
+)
+
+// NewCertificate creates a Certificate and ensures that the activity properly set empty
```
Isn't there a way in go to define private functions? If so, why don't split the constructor into a function `createCertificate()` as well as `assertActivityNotEmpty()` ?

aweisser replied:
Everything constant, method, function or variable that starts with lowercase is not exported but package private.
Everything that starts with upper case is exported and can be used from outside the package.

davelosert commented on domain/cert/certificate.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// OwnerID returns a copy of the owner ID,
+// so that the private attribute can't be manipluated from outside the package.
```
I'm not too familiar with Go, but isn't the way this is implemented making it clear that a copy is returned and hence this comment may also be unnecessary?

davelosert reviewed (changes requested):
@aweisser : Code-Wise, I think it's solid. One duplicated test that needs to be removed - and a lot of (imho) unnecessary comments that can be avoided.

aweisser committed:
Merge pull request #4 from aweisser/feature/Create_Make_File

Feature/create make file

aweisser committed:
Comments are not mandatory any more

The Go linter marks missing comments of exported types as a warning.
This often leads to obvious comments that are not helpful.
We'll rather stick to a common convention to leave helpfull comments
wherever needed and don't comment the obvious.

aweisser committed:
Comments are not mandatory any more

The Go linter marks missing comments of exported types as a warning.
This often leads to obvious comments that are not helpful.
We'll rather stick to a common convention to leave helpfull comments
wherever needed and don't comment the obvious.

aweisser committed:
Solve merge conflicts

aweisser committed:
Comments are not mandatory any more

aweisser committed:
Fix misspellings

aweisser commented:
@davelosert pls review
